test/libsolidity/smtCheckerTests/inheritance/base_contract_assertion_fail_2.sol

```diff
@@ -0,0 +1,25 @@
+pragma experimental SMTChecker;
+
+contract A {
+	uint x;
+	function f() internal view {
+		assert(x == 0);
+	}
+}
+
+contract B is A {
+	uint a;
+	uint b;
+}
+
+contract C is B {
+	uint y;
+	uint z;
+	uint w;
```

Comment:
@cameel the bug was that the cex generator sees `C.f()` (which exists via inheritance) but uses only the state variables from `A` because the AST scope of `f` is `A`. So if the derived contracts did not have more state variables it was fine.

Reply:
So now, instead of getting the scope where function `f` is defined (which is contract `A`) it will take the contract it is currently analyzing (which is `C`)? Yeah, I see how this fixes things.

Does this also affect virtual calls in any way? I.e. if contract `B` is being analyzed and the `assert()` in `f()` happens to call a hypothetical virtual function `v()` defined in `A` and later overridden in `C`? My guess is that it might be a problem because you'd be assuming contract `B` as context and not taking into account state variables from `C`. But maybe that's already handled in some way?

Reply:
If `B` is being analyzed and the inheritance structure is the same as in this example, `C.v()` doesn't exist yet, and a call to `v()` will either be `B.v()` or `A.v()`. When `C` is being analyzed, its `f()` is actually `B.f()`, and a call to `v()` in `B` will correctly call `C.v()`. tl;dr: the encoding is correct with respect to the calls themselves, the bug was only in how the cex generator saw the function predicates. I'll add a test similar to what you described.

Reply:
> `C.v()` doesn't exist yet, and a call to `v()` will either be `B.v()` or `A.v()`

Oh, you're right. It can't call `C.v()` when we're analyzing `B`.